feat(article): Scrollspy for the table of contents

utybo opened the pull request:
A bunch of things in this PR. I'll go ahead and say right away: feel free to *not* merge this if you don't feel like this is a good change. This is something I wanted to make for my own website, and I felt like it might be a good idea to upstream these changes.

**Changes:**

- Added the Hugo build lock file to the gitignore. This is a [file added by new Hugo versions that needs to be gitignore'd](https://gohugo.io/news/0.89.0-relnotes/#notes)
- Added ES2015 as the target in `jsconfig.json`. VS Code was reporting issues from existing TS code because this was not set.
- Removed the existing `scroll-behavior: smooth` and use a custom smoothing system (`smoothAnchors.ts`). This is required because Chromium does not support smooth scrolling on both the main page via an `a` link and on another element via a `scrollTo()` call.
- Added a CSS rule for the scrollbar thumb in the table of contents. There was actually one before, but it was the same color as the background, making it invisible. This is now fixed.
- Added a scroll spy system that displays the currently viewed section within the table of contents and automatically scrolls to display said section.

Here's an example of it in action:

https://user-images.githubusercontent.com/5175705/143720393-22946e89-ed30-4e4c-ab6b-b1db34d741b5.mp4




**TODO:**

- [x] Offset for each section needs to be recomputed when the window size changes
- [x] Video demo
- [x] Support `ol` and not just `ul`. ~~Seems that the approach I have used styling-wise does not work with the custom `ol` elements being used.~~ Actually works fine but required a modification on how ordered lists work in ToCs

CaiJimmy commented:
Thank you so much for the PR, it looks really nice 👍.

Sorry to say it but I might take a long time to review this due to the lack of time (and I don't even have the environment set up on my laptop). 😓

CaiJimmy commented on assets/scss/partials/base.scss:
```diff
@@ -1,7 +1,6 @@
 html {
     font-size: 62.5%;
     overflow-y: scroll;
-    scroll-behavior: smooth;
```
> because Chromium does not support smooth scrolling on both the main page via an a link and on another element via a scrollTo() call.

I re-added this line and removed `smoothAnchors.ts`, tried on Chromium 99.0.4839.0 on Windows 11, and don't see the issue (smooth scroll still working if I click on TOC link). Can you give me some information about this?

utybo replied:
Will recheck, it might have been patched

utybo replied:
Okay, seems like this is still the issue. To reproduce, you can take a footnote link in a long document, the animation will glitch when clicking on the footnote link and "snap" to the next section because the animation in the ToC stops the ongoing scroll animation in the main document.

Here's a video illustrating this:

https://user-images.githubusercontent.com/5175705/150031692-42cf2234-392f-4611-a960-3eefdea9e97b.mp4

utybo replied:
Here are some links related to this issue. I will also add them in the `smoothAnchors` file to properly document the reason for that file's existence.

- https://stackoverflow.com/questions/49318497/google-chrome-simultaneously-smooth-scrollintoview-with-more-elements-doesn
- https://stackoverflow.com/questions/57214373/scrollintoview-using-smooth-function-on-multiple-elements-in-chrome
- https://bugs.chromium.org/p/chromium/issues/detail?id=833617
- https://bugs.chromium.org/p/chromium/issues/detail?id=1043933
- https://bugs.chromium.org/p/chromium/issues/detail?id=1121151

CaiJimmy replied:
I see, thanks for the information!

CaiJimmy commented on assets/ts/scrollspy.ts:
```diff
@@ -0,0 +1,125 @@
+        // - No active section but the link does not exist in toc (e.g. because it is outside of the applicable ToC levels) => undefined
+        let newActiveSectionLink: HTMLElement | undefined 
+        if (newActiveSection) {
+            newActiveSectionLink = findLinkForSectionId(newActiveSection.id, navigation);
```
Instead of iterating HTML each time, since the TOC element does not change dynamically, it might be more efficient to build a map (id -> link ref) outside `scrollHandler`:

```js
const sectionLinkRef: { [key: string]: HTMLElement } = {};
navigation.forEach((navigationElement: HTMLElement) => {
    const link = navigationElement.querySelector("a");
    const id = link.getAttribute("href").replace('#', '');
    sectionLinkRef[id] = navigationElement;
});
```

And accessing it by:
```typescript
if (newActiveSection) {
    /// Undefined if id does not exist
    newActiveSectionLink = sectionLinkRef[newActiveSection.id];
}
```

How does that sound to you?

utybo replied:
Good idea, I didn't think of that 👍🏻

utybo replied:
Done, let me know what you think. I didn't go with the exact same code you provided, but it does the same thing.

CaiJimmy commented:
There's a merge conflict with `jsconfig.json` because I've modified it in #447. Instead of ES2015 I added ES2020 as target because I used `String.matchAll` method.

Since I don't have write permission to your branch, I can't perform a merge commit.

utybo commented:
> There's a merge conflict with `jsconfig.json` because I've modified it in #447. Instead of ES2015 I added ES2020 as target because I used `String.matchAll` method.
> 
> Since I don't have write permission to your branch, I can't perform a merge commit.

I've done the merge commit :+1: Sorry for the write permission issue, I prefer to leave that option turned off

CaiJimmy reviewed (approved):
LGTM!

Thanks ⭐
